Port couple of FxCop rules (CA1810 and CA2207)

mavasani opened the pull request:
Port FxCop rules CA1810 (Initialize reference type static fields inline) and CA2207 (Initialize value type static fields inline).

NOTE: I have merged both the rules into a single analyzer as the rule just differs by the enclosing type kind (struct versus class)

mavasani commented:
@srivatsn @basoundr @michaelcfanning @heejaechang @jinujoseph @lgolding @natidea

heejaechang commented on src/System.Runtime.Analyzers/Core/InitializeStaticFieldsInline.Fixer.cs:
```diff
@@ -27,10 +29,10 @@ public sealed override FixAllProvider GetFixAllProvider()
+        {
+            // TODO: Implement the fixer.
             // This is to get rid of warning CS1998, please remove when implementing this analyzer
             await new Task(() => { });
```
you can remove "async"

mavasani replied:
I am planning to delete the fixer file completely, not sure if it is worth writing a fixer for this.

srivatsn replied:
The fix would be a bit involved because it would need to find all static fields being assigned in the constructor and make them all as initializers. It is a mechanical transformation and saves a bunch of copy\pasting - sounds useful actually.

mavasani commented:
@genlu

genlu commented on src/System.Runtime.Analyzers/UnitTests/InitializeValueTypeStaticFieldsInlineTests.cs:
```diff
@@ -12,12 +12,328 @@ public class InitializeValueTypeStaticFieldsInlineTests : DiagnosticAnalyzerTest
+End Structure",
+    GetCA2207BasicDefaultResultAt(4, 13, "Struct1"));
+
+        }
```
Are those tests copied from FxCop, they are really nice :)

mavasani replied:
No, I generally use the FxCop tests for reference, but didn't do so this time :)

michaelcfanning commented:
Note misspelling in NoDuplicteDiagnostics.

genlu commented:
The analyzer LGTM. Please delete or implement CodeFix.

lgolding commented on src/System.Runtime.Analyzers/UnitTests/InitializeValueTypeStaticFieldsInlineTests.cs:
```diff
@@ -12,12 +12,328 @@ public class InitializeValueTypeStaticFieldsInlineTests : DiagnosticAnalyzerTest
+        }
+
+        [Fact]
+        public void Basic_CA1810_Diagnostic_NoDuplicteDiagnostics()
```
"Duplicate"

michaelcfanning commented on src/System.Runtime.Analyzers/UnitTests/InitializeValueTypeStaticFieldsInlineTests.cs:
```diff
@@ -12,12 +12,328 @@ public class InitializeValueTypeStaticFieldsInlineTests : DiagnosticAnalyzerTest
+        }
+
+        [Fact]
+        public void CSharp_CA1810_NoDiagnostic_NoStaticFieldInitializedInStaticConstructor()
```
@lgolding @genlu  we need a unit test naming standard to be declared.

lgolding replied:
I actually like this. It sorts nicely by language and by diagnostic ID, and it separates positive from negative tests. As fond as I am of "behavioral naming" ("it should do so-and-so"), I think this convention works better for analyzers. Because for analyzers the behavior will always be either "it emits a diagnostic" or "it doesn't".

michaelcfanning commented on src/System.Runtime.Analyzers/VisualBasic/System.Runtime.VisualBasic.Analyzers.vbproj:
```diff
@@ -116,8 +116,8 @@
     <Compile Include="BasicDoNotRaiseReservedExceptionTypes.Fixer.vb" />
-    <Compile Include="BasicInitializeValueTypeStaticFieldsInline.vb" />
-    <Compile Include="BasicInitializeValueTypeStaticFieldsInline.Fixer.vb" />
+    <Compile Include="BasicInitializeStaticFieldsInline.vb" />
```
@lgolding @genlu  we need a nice text naming convention defined as well. It is pleasant when the test names map to the rule names (which map to the rule file names)

lgolding replied:
In general I agree. I don't think we want every test method to have (for example) <RuleName> (_e.g._, `InitializeStaticFieldsInline` as part of its name. It's too verbose. But I _do_ like having it as part of the file name. So how about

* File name is <RuleId>_<Language><RuleName>.<ext> for the file name, for example `CA9999_BasicInitializeStaticFieldsInline.vb`
* Continue to mention only the <RuleId> in the test method names
* Continue to name the classes with "<Language><RuleName>" (e.g. `class BasicInitializeStaticFieldsInline`).

michaelcfanning replied:
+1. I think we can trim the 'Basic' prefix, it's confusing in any case for these C# tests. I'd suggest naming these classes [RuleName]Tests.cs|vb and [RuleName]FixerTests

lgolding replied:
If you trim the Language portion from the class name, then the C# and VB classes will have the same name. (You will frequently have C# and VB implementations of the same rule, derived from a common abstract base class, thus:

Abstract class Foo in Foo.cs
Concrete derived class CSharpFoo in CSharpFoo.cs (or CA9999_CSharpFoo.cs as proposed above)
Concrete derived class BasicFoo in BasicFoo.vb (or CA9999_BasicFoo.vb)

I don't see how you can trim the language prefix.

mavasani replied:
I don't think we want more than one unit test projects per analyzer - so we don't want a VB project for testing VB implementation of the analyzer + C# project for testing C# implementation. How about the following:

1. Retain Gen's current project structure - have just a single C# unit test project per analyzer.
2. Create a single [RuleName]Tests.cs file for all rule tests.
3. For analyzers that implement a single CA rule (and most analyzers will do so), we should name the tests without the language or CA rule prefix and put equivalent C# and VB tests in the same unit test - this is what Gen did for his PR (see [here](https://github.com/dotnet/roslyn-analyzers/pull/340/files#r44174198)), and I like that approach.
4. This analyzer merges 2 rules, so its probably okay for the tests for it to have CAXXXX prefix OR create separate nested types for each rule tests - its just a minor coding style preference which applies only for a minority of rules.

lgolding replied:
I think there's some confusion. I don't think anybody suggested having more than one "test project" (did you mean "test file"?) per analyzer.

mavasani replied:
I thought this comment suggested having C# and VB files for corresponding analyzer implementations:

```
Concrete derived class CSharpFoo in CSharpFoo.cs (or CA9999_CSharpFoo.cs as proposed above)
Concrete derived class BasicFoo in BasicFoo.vb (or CA9999_BasicFoo.vb)
```

genlu replied:
I think naming convention and style is an issue we must agree on across all the analyzer implementations, thus not suitable to discuss here in a PR. I have create issue https://github.com/dotnet/roslyn-analyzers/issues/347 for this, please have all the discussion over there, and hopefully we can get a finalized documentation on this.

@michaelcfanning @lgolding @srivatsn @mavasani

mavasani commented:
@lgolding @michaelcfanning @genlu I'll address the feedback on naming for the test class/methods, typo on Duplicate and deleting the unused fixer file in the next commit. Thanks for reviewing!

srivatsn commented on src/System.Runtime.Analyzers/CSharp/CSharpInitializeStaticFieldsInline.cs:
```diff
@@ -0,0 +1,24 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class CSharpInitializeStaticFieldsInlineAnalyzer : InitializeStaticFieldsInlineAnalyzer<SyntaxKind>
+    {
+        protected override SyntaxKind AssignmentNodeKind => SyntaxKind.SimpleAssignmentExpression;
```
What about other kinds of assignment like:
- a, b = 10
- a += 10 (and all the other compound assignments)
- Foo(ref\out a)

mavasani replied:
We will automatically handle these cases when we port this to IOperation based analyzer.

srivatsn replied:
We should create a bug to do that then. I am not sure if the default stance would be port every analyzer to IOperation

mavasani replied:
I think we need to do this exercise for all ported rules when we move to Roslyn 1.2  - I will create a work item.

srivatsn commented:
You also need to update the props file. (Yes Manish, we need that analyzer to check if the props file matches rules).

mavasani commented:
I have addressed the feedback - please take a look at the unit tests again.

@genlu I have left the fixer as unimplemented. As Sri recommends, we probably want to implement it in future.

genlu commented:
Could you please also mark the rules as ported in the spreadsheet?

LGTM

mavasani commented:
Done, thanks Gen.
